### PR TITLE
[*] TEST : Unit test tooling

### PR DESCRIPTION
**Pull request by mdomenjoud**
First set of tools to ease Unit testing : 
- static singleton setters for testing.
- preloaded mocks for common static singletons.
- common classes for integration and unit tests.

### CODE REVIEW

**Comment by SebSept** [comment]
Hi,

Prestashop needs automated tests, I agree.

Do you plan to write mocks for each components which has static methods ? 
When the components change, you have to rewrite the mocks, that's a big work, and it introduce new inconsistencies possibility.
Am I wrong ?

**Comment by mdomenjoud** [comment]
Hello,
> Do you plan to write mocks for each components which has static methods ? 

We plan to enable unit testing by gradually improving design. One solution is to expose these mocks, even if this is not the best to break dependencies in a clean way : in the end we'd like to get rid off these static components.

> When the components change, you have to rewrite the mocks, that's a big work, and it introduce new inconsistencies possibility.
> Am I wrong ?

As we plan to gradually test all changes, a change on a component will need new tests, so the mocks would of course have to change if their exposed functions change : if should be fine, as soon as tests rely on components behaviour and not implementation, and test code is maintained as first class citizen.

**Comment by SebSept** [comment]
Ok thanks for answer.

File: classes/Context.php [hunk 1]
```diff
@@ -270,6 +270,23 @@ public static function getContext()
 		return self::$instance;
 	}
 
+    /**
+     * @param $testInstance Context
+     * Unit testing purpose only
+     */
+    public static function setInstanceForTesting($testInstance)
```
Comment (jeromenadaud) [inline classes/Context.php]: norms

File: classes/cache/Cache.php [hunk 2]
```diff
@@ -134,6 +134,23 @@ public static function getInstance()
 		return self::$instance;
 	}
 
+    /**
+     * @param $testInstance Cache
+     * Unit testing purpose only
+     */
+    public static function setInstanceForTesting($testInstance)
```
Comment (jeromenadaud) [inline classes/cache/Cache.php]: norms

File: classes/db/Db.php [hunk 3]
```diff
@@ -208,7 +208,10 @@ public static function getInstance($master = true)
 				array('server' => _DB_SERVER_, 'user' => _DB_USER_, 'password' => _DB_PASSWD_, 'database' => _DB_NAME_), /* MySQL Master server */
 			);
 
-		Db::loadSlaveServers();
+		if (!$master)
+		{
+			Db::loadSlaveServers();
```
Comment (jeromenadaud) [inline classes/db/Db.php]: norms

File: classes/db/Db.php [hunk 4]
```diff
@@ -233,6 +236,23 @@ public static function getInstance($master = true)
 		return self::$instance[$id_server];
 	}
 
+    /**
+     * @param Db
+     * Unit testing purpose only
+     */
+    public static function setInstanceForTesting($testDatabaseManager)
```
Comment (jeromenadaud) [inline classes/db/Db.php]: norms

**Comment by SebSept** [comment]
using the `setInstanceForTesting($testInstance)` looks like a good idea.

With this sort of framework how can you set what a mocked instance will return ?

File: classes/db/Db.php [hunk 5]
```diff
@@ -63,10 +63,12 @@
...
 
 	/** @var array List of DB instances */
-	protected static $instance = array();
+	public static $instance = array();
 
 	/** @var array List of server settings */
-	protected static $_servers = array();
+	public static $_servers = array();
+
+	public static $_slave_servers_loaded = null;
```
Comment (jeromenadaud) [inline classes/db/Db.php]: Missing PHP DOC ;-)

**Pull request merged**